src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt

```diff
@@ -148,4 +161,27 @@ class UserService(
 
         return tokens
     }
+
+    override suspend fun updateUser(request: UserOuterClass.User.Update): UserOuterClass.User {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+
+        // When the email should be changed, then check whether a user with the given email already exists
+        if (request.mask.pathsList.contains("email") && userRepo.doesUserExistByEmail(request.user.email)) {
+            throw DuplicateEntityException.UserEmail(request.user.email)
+        }
+
+        // When the role should be changed, then check whether the requesting user a server admin
+        if (request.mask.pathsList.contains("role") && currentUser.role != UserRole.USER_ROLE_ADMIN) {
+            throw UnauthorizedException.Single.User(currentUser.id.toString(), request.user.id, AccessType.UPDATE)
+        }
+
+        // When user information (except for the role) should be changed, then check whether the requesting user changes
+        // its own information or changes to information from another user (except the user is an admin)
+        if (currentUser.id.toString() != request.user.id && currentUser.role != UserRole.USER_ROLE_ADMIN) {
+            throw UnauthorizedException.Single.User(currentUser.id.toString(), request.user.id, AccessType.UPDATE)
+        }
```

Comment:
Here you can use the helper too. But you can also combine both checks, as they do the same.

```suggestion
        // Check whether the current user is a server admin if the role is changed or the requested user is different
        // from the current user
        if (request.mask.pathsList.contains("role") || currentUser.id != requestedUser.id) {
            verifyServerAdminRole(currentUser) { UnauthorizedException.Single.User(it, request.user.id, AccessType.UPDATE) }
        }
```

Reply:
I combined them.